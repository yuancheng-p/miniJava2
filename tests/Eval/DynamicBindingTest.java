public class DynamicBindingTest {

  public static void main(String [] args) {
    A b = new B();
    b.foo();
    b = new C();
    b.foo();
    b = new A();
    b.foo();
  }
}

class A {
  public int foo () {
    return 100;
  }
}


class B extends A {
  public int foo () {
    return 200;
  }
}


class C extends A {
  public int foo () {
    return 300;
  }
}
