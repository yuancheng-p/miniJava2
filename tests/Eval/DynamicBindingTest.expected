successfull parsing
start typing
check add and methods:DynamicBindingTest:main
check add and methods:A:foo
check add and methods:B:foo
check add and methods:C:foo
--------- env --------
A : class_env(parent=: Object)
method : foo
B : class_env(parent=: A)
method : foo
Boolean : class_env(parent=: Object)
C : class_env(parent=: A)
method : foo
DynamicBindingTest : class_env(parent=: Object)
method : main paramlist: String[] 
Integer : class_env(parent=: Object)
Object : class_env(parent=: )
String : class_env(parent=: Object)
end typing
_______________main___________________
A b = new B();
    /*
    ------ b in frame ------
    b : 0 (ref)
    ------ object in heap ------
    [B] (0) :
    */

b.foo();
    /*
    ------ Call method: foo() ---------
    ------ Return: 200 ------------
    */

b=new C();
b.foo();
    /*
    ------ Call method: foo() ---------
    ------ Return: 300 ------------
    */

b=new A();
b.foo();
    /*
    ------ Call method: foo() ---------
    ------ Return: 100 ------------
    */

